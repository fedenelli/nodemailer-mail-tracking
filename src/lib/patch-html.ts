import { sign, JwtOptions } from './jwt';
import { JwtData, JwtDataForLink, MailTrackOptions } from '../types';

export const addBlankImage = (
  options: Pick<MailTrackOptions, 'baseUrl'> & JwtOptions,
  html: string,
  data: JwtData
) => {
  const jwtImg = sign(options, data);
  const imgTag = `<img src="${options.baseUrl}/blank-image/${jwtImg}" />`;
  if (new RegExp(/<\/body>/).test(html)) {
    html = html.replace(/<\/body>/, `${imgTag}</body>`);
  } else {
    html = `${html}${imgTag}`;
  }

  return html;
};

const linkRE = /(<a\s+(?:[^>]*?\s+)?href=")([^"]*)(")/gi;
export const patchLinks = (
  options: Pick<MailTrackOptions, 'baseUrl'> & JwtOptions,
  html: string,
  data: JwtData
) => {
  return html.replace(linkRE, (_, first, link, end) => {
    if (link.match(/^#/)) {
      return [first, link, end].join('');
    }
    const jwtData: JwtDataForLink = {
      ...data,
      link,
    };
    const jwtLink = sign(options, jwtData);
    return `${first}${options.baseUrl}/link/${jwtLink}${end}`;
  });
};
