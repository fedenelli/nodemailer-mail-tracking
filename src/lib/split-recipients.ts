import {
  SendMailOptions,
  SendMailOptionsWithHtml,
  SendMailOptionsPatched,
  IRecipient,
} from '../types';

export const splitByRecipients = (
  options: SendMailOptionsWithHtml
): SendMailOptionsPatched[] => {
  const recipientKeys = ['to', 'cc', 'bcc'];
  const results: SendMailOptionsPatched[] = [];

  for (let recipientKey of recipientKeys) {
    if (options[recipientKey as keyof typeof options]) {
      results.push(
        patchSendMailOptions(
          options[recipientKey as keyof typeof options] as IRecipient,
          options
        )
      );
    }
  }
  return results;
};

const patchSendMailOptions = (
  recipient: IRecipient,
  { to, cc, bcc, ...options }: SendMailOptionsWithHtml
): SendMailOptionsPatched => {
  const headers: SendMailOptions['headers'] = {};
  const toAddress = getRecipientAddress(to);
  if (toAddress) {
    headers.To = toAddress;
  }
  const ccAddress = getRecipientAddress(cc);
  if (ccAddress) {
    headers.Cc = ccAddress;
  }
  const bccAddress = getRecipientAddress(bcc);
  if (bccAddress) {
    headers.Bcc = bccAddress;
  }

  const from = getRecipientAddress(options.from);
  const rcptTo = getRecipientAddress(recipient);

  return {
    ...options,
    headers: {
      ...options.headers,
      ...headers,
    },
    envelope: {
      from: (Array.isArray(from) ? from[0] : from) || false,
      to: Array.isArray(rcptTo) ? rcptTo : [rcptTo as string],
    },
  };
};

function getRecipientAddress(
  recipient: IRecipient
): string | string[] | undefined {
  return typeof recipient === 'string'
    ? recipient
    : Array.isArray(recipient)
    ? recipient
        .map(r => {
          return r === 'string'
            ? r
            : ((r as any).address as string) || undefined;
        })
        .filter(<T>(n?: T): n is T => Boolean(n))
    : recipient?.address
    ? recipient.address
    : undefined;
}
