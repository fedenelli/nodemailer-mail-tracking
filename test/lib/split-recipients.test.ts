import { splitByRecipients } from '../../src/lib/split-recipients';

describe('splitByRecipients', () => {
  it('works', () => {
    expect(
      splitByRecipients({
        subject: 'Test',
        from: 'sender@email.com',
        to: 'some@email.com',
        cc: 'cc@email.com',
        bcc: 'cc@email.com',
        replyTo: 'reply@email.com',
        html: 'Hey!',
        text: 'bim',
      })
    ).toMatchInlineSnapshot(`
      Array [
        Object {
          "envelope": Object {
            "from": "sender@email.com",
            "to": "some@email.com",
          },
          "from": "sender@email.com",
          "headers": Object {
            "Bcc": "cc@email.com",
            "Cc": "cc@email.com",
            "To": "some@email.com",
          },
          "html": "Hey!",
          "replyTo": "reply@email.com",
          "subject": "Test",
          "text": "bim",
        },
        Object {
          "envelope": Object {
            "from": "sender@email.com",
            "to": "cc@email.com",
          },
          "from": "sender@email.com",
          "headers": Object {
            "Bcc": "cc@email.com",
            "Cc": "cc@email.com",
            "To": "some@email.com",
          },
          "html": "Hey!",
          "replyTo": "reply@email.com",
          "subject": "Test",
          "text": "bim",
        },
        Object {
          "envelope": Object {
            "from": "sender@email.com",
            "to": "cc@email.com",
          },
          "from": "sender@email.com",
          "headers": Object {
            "Bcc": "cc@email.com",
            "Cc": "cc@email.com",
            "To": "some@email.com",
          },
          "html": "Hey!",
          "replyTo": "reply@email.com",
          "subject": "Test",
          "text": "bim",
        },
      ]
    `);
  });

  const attrs = ['to', 'cc', 'bcc'];

  attrs.forEach(attr => {
    it(`works with 1 ${attr}`, () => {
      expect(
        splitByRecipients({
          from: 'sender@email.com',
          [attr]: 'some@email.com',
          html: 'Hey!',
        })
      ).toMatchSnapshot();
    });
    it(`works with 1 multiple ${attr}`, () => {
      expect(
        splitByRecipients({
          from: 'sender@email.com',
          [attr]: ['some@email.com'],
          html: 'Hey!',
        })
      ).toMatchSnapshot();
    });
    it(`works with multiple ${attr}`, () => {
      expect(
        splitByRecipients({
          from: 'sender@email.com',
          [attr]: ['some@email.com', 'other@mail.com'],
          html: 'Hey!',
        })
      ).toMatchSnapshot();
    });
  });
});
