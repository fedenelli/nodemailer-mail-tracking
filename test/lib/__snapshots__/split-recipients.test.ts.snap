// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`splitByRecipients works with 1 bcc 1`] = `
Array [
  Object {
    "envelope": Object {
      "from": "sender@email.com",
      "to": "some@email.com",
    },
    "from": "sender@email.com",
    "headers": Object {
      "Bcc": "some@email.com",
    },
    "html": "Hey!",
  },
]
`;

exports[`splitByRecipients works with 1 cc 1`] = `
Array [
  Object {
    "envelope": Object {
      "from": "sender@email.com",
      "to": "some@email.com",
    },
    "from": "sender@email.com",
    "headers": Object {
      "Cc": "some@email.com",
    },
    "html": "Hey!",
  },
]
`;

exports[`splitByRecipients works with 1 multiple bcc (with name) 1`] = `
Array [
  Object {
    "envelope": Object {
      "from": "sender@email.com",
      "to": "Some Email <some@email.com>",
    },
    "from": "sender@email.com",
    "headers": Object {
      "Bcc": Array [
        "Some Email <some@email.com>",
      ],
    },
    "html": "Hey!",
  },
]
`;

exports[`splitByRecipients works with 1 multiple bcc 1`] = `
Array [
  Object {
    "envelope": Object {
      "from": "sender@email.com",
      "to": "some@email.com",
    },
    "from": "sender@email.com",
    "headers": Object {
      "Bcc": Array [
        "some@email.com",
      ],
    },
    "html": "Hey!",
  },
]
`;

exports[`splitByRecipients works with 1 multiple cc (with name) 1`] = `
Array [
  Object {
    "envelope": Object {
      "from": "sender@email.com",
      "to": "Some Email <some@email.com>",
    },
    "from": "sender@email.com",
    "headers": Object {
      "Cc": Array [
        "Some Email <some@email.com>",
      ],
    },
    "html": "Hey!",
  },
]
`;

exports[`splitByRecipients works with 1 multiple cc 1`] = `
Array [
  Object {
    "envelope": Object {
      "from": "sender@email.com",
      "to": "some@email.com",
    },
    "from": "sender@email.com",
    "headers": Object {
      "Cc": Array [
        "some@email.com",
      ],
    },
    "html": "Hey!",
  },
]
`;

exports[`splitByRecipients works with 1 multiple to (with name) 1`] = `
Array [
  Object {
    "envelope": Object {
      "from": "sender@email.com",
      "to": "Some Email <some@email.com>",
    },
    "from": "sender@email.com",
    "headers": Object {
      "To": Array [
        "Some Email <some@email.com>",
      ],
    },
    "html": "Hey!",
  },
]
`;

exports[`splitByRecipients works with 1 multiple to 1`] = `
Array [
  Object {
    "envelope": Object {
      "from": "sender@email.com",
      "to": "some@email.com",
    },
    "from": "sender@email.com",
    "headers": Object {
      "To": Array [
        "some@email.com",
      ],
    },
    "html": "Hey!",
  },
]
`;

exports[`splitByRecipients works with 1 to 1`] = `
Array [
  Object {
    "envelope": Object {
      "from": "sender@email.com",
      "to": "some@email.com",
    },
    "from": "sender@email.com",
    "headers": Object {
      "To": "some@email.com",
    },
    "html": "Hey!",
  },
]
`;

exports[`splitByRecipients works with multiple bcc 1`] = `
Array [
  Object {
    "envelope": Object {
      "from": "sender@email.com",
      "to": "some@email.com",
    },
    "from": "sender@email.com",
    "headers": Object {
      "Bcc": Array [
        "some@email.com",
        "other@mail.com",
      ],
    },
    "html": "Hey!",
  },
  Object {
    "envelope": Object {
      "from": "sender@email.com",
      "to": "other@mail.com",
    },
    "from": "sender@email.com",
    "headers": Object {
      "Bcc": Array [
        "some@email.com",
        "other@mail.com",
      ],
    },
    "html": "Hey!",
  },
]
`;

exports[`splitByRecipients works with multiple cc 1`] = `
Array [
  Object {
    "envelope": Object {
      "from": "sender@email.com",
      "to": "some@email.com",
    },
    "from": "sender@email.com",
    "headers": Object {
      "Cc": Array [
        "some@email.com",
        "other@mail.com",
      ],
    },
    "html": "Hey!",
  },
  Object {
    "envelope": Object {
      "from": "sender@email.com",
      "to": "other@mail.com",
    },
    "from": "sender@email.com",
    "headers": Object {
      "Cc": Array [
        "some@email.com",
        "other@mail.com",
      ],
    },
    "html": "Hey!",
  },
]
`;

exports[`splitByRecipients works with multiple to 1`] = `
Array [
  Object {
    "envelope": Object {
      "from": "sender@email.com",
      "to": "some@email.com",
    },
    "from": "sender@email.com",
    "headers": Object {
      "To": Array [
        "some@email.com",
        "other@mail.com",
      ],
    },
    "html": "Hey!",
  },
  Object {
    "envelope": Object {
      "from": "sender@email.com",
      "to": "other@mail.com",
    },
    "from": "sender@email.com",
    "headers": Object {
      "To": Array [
        "some@email.com",
        "other@mail.com",
      ],
    },
    "html": "Hey!",
  },
]
`;
